### PR TITLE
add a HasSynced() for DeltaFIFO and FIFO, method, which is very helpful for Informer

### PR DESCRIPTION
**Pull request by mqliang**
ref https://github.com/kubernetes/kubernetes/issues/20259

provide a `HasSynced()` method, which is very helpful for `Informer`

### CODE REVIEW

**Comment by mqliang** [comment]
Has not add unit test yet. Send for review and feedback, hoping for finding some bugs or more elegant implementation.

**Comment by mqliang** [comment]
@kargakis you may interested.

**Comment by lavalamp** [comment]
I think I see what you're trying to do. However, I believe this change makes the following part of the documentation for this object wrong:

> // * When you process an object, you want to see everything
> //    that's happened to it since you last processed it.

Since the replace goes to the first queue, and other changes go to the other queue. This makes me nervous. I'm also a bit nervous about requiring a Replace before anything else can be done.

As an alternative suggestion, how about
* Keep a synced flag. Initially, this is false.
* If an add, update, or delete are called first, then set the flag to true.
* If an update is called first, set a counter to be the length of the queue. Decrement this counter with every Pop(), when it reaches 0, set the synced flag to true.

What do you think about that?

**Comment by mqliang** [comment]
@lavalamp reworked. PTAL

**Comment by mqliang** [comment]
@k8s-bot test this please

**Comment by mqliang** [comment]
> FAIL	k8s.io/kubernetes/pkg/controller/framework	3.467s

I am confused since the unit test could pass on me PC:

>/usr/local/go/bin/go test -v k8s.io/kubernetes/pkg/controller/framework -run "^Example|ExampleInformer|TestHammerController|TestUpdate$"
Testing started at 8:18 PM ...
	controller_test.go:283: got: map[string][]string{}PASS
ok  	k8s.io/kubernetes/pkg/controller/framework	1.255s

**Comment by mqliang** [comment]
@lavalamp bug fixed. PTAL

File: pkg/client/cache/delta_fifo.go [hunk 1]
```diff
@@ -100,6 +100,11 @@ type DeltaFIFO struct {
 	items map[string]Deltas
 	queue []string
 
+	// firstCalled is used to record the first method called by caller
+	firstCalled int
+	// count is the number of items inserted by the first call of Replace()
+	count int
```
Comment (lavalamp) [inline pkg/client/cache/delta_fifo.go]: May I suggest, instead:

```go
populated bool
initialPopulationCount int
```

Then all of add/update/delete set populated to true; and replace does

```go
if !f.populated {
  f.populated = true
  f.initialPopulationCount = len(list)
}
```

HasSynced does

```go
return f.populated && f.initialPopulationCount == 0
```

I think this is simpler than keeping track of the exact method first called, I'd rather not do that unless needed.

**Comment by lavalamp** [comment]
I like this approach a lot better than the first version, thanks!

**Comment by mqliang** [comment]
@lavalamp Reworked. PTAL

**Comment by mqliang** [comment]
@lavalamp Updated. PTAL.

**Comment by lavalamp** [comment]
LGTM, very clean fix. Thanks!

**Comment by mqliang** [comment]
@lavalamp Should we cherry pick this to k8s v1.0(and maybe v1.1) ?

**Comment by lavalamp** [comment]
@mqliang IMO, this isn't a big enough problem, but if you think it's super urgent we can get a second opinion.

File: pkg/client/cache/delta_fifo.go [hunk 2]
```diff
@@ -373,6 +394,12 @@ func (f *DeltaFIFO) Replace(list []interface{}, resourceVersion string) error {
 	f.lock.Lock()
 	defer f.lock.Unlock()
 	keys := make(sets.String, len(list))
+
+	if !f.populated {
+		f.populated = true
+		f.initialPopulationCount = len(list)
```
Comment (timothysc) [inline pkg/client/cache/delta_fifo.go]: Don't you want to set this even if populated?

**Pull request merged**